plugins/outputs/nsca/nsca.go

```diff
@@ -0,0 +1,167 @@
+	Timeout time.Duration
+}
+
+// Message is the contents of an NSCA message
+type message struct {
+	// State is one of {STATE_OK, STATE_WARNING, STATE_CRITICAL, STATE_UNKNOWN}
+	state string
+	// Host is the host name to set for the NSCA-ng message
+	host string
+	// Service is the service name to set for the NSCA-ng message
+	service string
+	// Message is the "plugin output" of the NSCA-ng message
+	message string
+}
+
+// NSCAServer can be used as a lower-level alternative to RunEndpoint. It is NOT safe
+// to use an instance across mutiple threads.
+type NSCAServer struct {
+	Url          string
+	conn         net.Conn
+	serializer   serializers.Serializer
+	serverInfo   ServerInfo
+	CPUWarning   float64
+	CPUCritical  float64
+	DiskWarning  float64
+	DiskCritical float64
+}
+
+var sampleConfig = `
+  Url = "icinga.dev.thermeon.com:5668"
+  ## Data format to output.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
+  ##data_format = "influx"
+  CPUWarning = 3.6
+  CPUCritical = 4.0
+  DiskWarning = 80.0
+  DiskCritical = 90.0
+`
+
+func (n *NSCAServer) SetSerializer(serializer serializers.Serializer) {
+	n.serializer = serializer
+}
+
+func (n *NSCAServer) SampleConfig() string {
+	return sampleConfig
+}
+
+func (n *NSCAServer) Description() string {
+	return "Send telegraf measurements to nsca"
+}
+
+// define GetKey and GetIdentity methods
+func getIdentity() string {
+	return "system-checker"
+}
+
+func getKey(id string) ([]byte, error) {
+	return []byte("0123456789"), nil
+}
+
+// create the appropriate TLS configuration
+// note that we specifiy a single cipher suite of type TLS_PSK_*
+// also note that the server requires a certificate, even if not used here
+
+var (
+	config = &tls.Config{
+		CipherSuites: []uint16{psk.TLS_PSK_WITH_AES_256_CBC_SHA},
+		Certificates: []tls.Certificate{tls.Certificate{}},
+		Extra: psk.PSKConfig{
+			GetKey:      getKey,
+			GetIdentity: getIdentity,
+		},
+	}
+)
+
+func generateSessionId() string {
+	randbytes := make([]byte, 6)
+	if _, err := io.ReadFull(rand.Reader, randbytes); err != nil {
+		return ""
+	}
+	return base64.StdEncoding.EncodeToString(randbytes)
+}
+
+// Connect to an NSCA server.
+func (n *NSCAServer) Connect() error {
+	conn, err := tls.Dial("tcp", n.Url, config)
+	if err != nil {
+		return err
+	}
+	n.conn = conn
+	return nil
+}
+
+// Close the connection and clean up.
+func (n *NSCAServer) Close() error {
+	if n.conn != nil {
+		n.conn.Close()
+		n.conn = nil
+	}
+	return nil
+}
+
+func (n *NSCAServer) Write(metrics []telegraf.Metric) error {
+
+	sessionId := generateSessionId()
+
+	//commands which nsca-ng server accepts
+	command := "MOIN 1 " + sessionId
+	fmt.Fprintf(n.conn, command+"\n")
+
+	var batch []byte
+	s, err := serializers.NewNscaSerializer(n.CPUWarning, n.CPUCritical, n.DiskWarning, n.DiskCritical)
```

Comment:
I wouldn't expect a new serializer each time. Should be part of n, surely?

Reply:
Yeah, placed in connect method. where it performs once.